pallene/coder.lua

```diff
@@ -1328,17 +1328,29 @@ end
 
 gen_cmd["SetUpvalue"] = function(self, cmd, _func)
     local func = self.module.functions[cmd.f_id]
-    local capture_upvalues = {}
 
-    local cclosure = string.format("clCvalue(&%s)", self:c_var(cmd.dst))
-    for i, upval_info in ipairs(func.captured_vars) do
-        local typ   = upval_info.decl.typ
-        local c_val = self:c_value(upval_info.value)
-        local upvalue_dst = string.format("&(%s->upvalue[%s])", cclosure, C.integer(i))
+    assert(cmd.src_f._tag == "ir.Value.LocalVar")
+    local cclosure = string.format("clCvalue(&%s)", self:c_var(cmd.src_f.id))
+
+    local capture_upvalues = {}
+    for i, val in ipairs(cmd.srcs) do
+        local typ   = func.captured_vars[i].typ
+        local c_val = self:c_value(val)
+        local upvalue_dst = string.format("&(ccl->upvalue[%s])", C.integer(i))
         table.insert(capture_upvalues, set_stack_slot(typ, upvalue_dst, c_val))
```

Comment:
I just noticed another thing: we're using `set_stack_slot` here even though it is actually a heap slot. We should have a comment explaining why it is OK.

The reason is that there are no GC calls between the `NewClosure` and the `SetUpvalue`. The newly created closure is guaranteed to be white and therefore the write barrier is not necessary.